File: packages/db/src/drivers/tag.ts
import { PgUpdateSetSource } from "drizzle-orm/pg-core";
import { db, drizzle, getGuildOrCreate, getUserOrCreate, schema } from "../";
import { and, eq } from "drizzle-orm";

export async function getTag(name: string, guildId: string) {
	return db.query.tags
		.findFirst({
			where: (table, { eq, and }) =>
				and(eq(table.name, name), eq(table.guildId, guildId)),
			with: {
				author: true,
			},
		})
		.execute();
}

export async function getTagOrThrow(
	name: string,
	guildId: string,
	cb: () => void,
) {
	const tag = await getTag(name, guildId);

	if (!tag) throw cb();

	return tag!;
}

export async function removeTag(name: string, guildId: string) {
	await getTag(name, guildId);

	const [tag] = await db
		.delete(schema.tags)
		.where(
			drizzle.and(
				drizzle.eq(schema.tags.name, name),
				drizzle.eq(schema.tags.guildId, guildId),
			),
		)
		.returning();

	return tag;
}

export async function createTag(
	name: string,
	content: string,
	userId: string,
	guildId: string,
) {
	await getUserOrCreate(userId);
	await getGuildOrCreate(guildId);

	const exists = await getTag(name, guildId);

	if (exists) throw new Error("This tag already exists on this guild");

	const [tag] = await db.transaction(async (tx) => {
		const [tag] = await tx
			.insert(schema.tags)
			.values({
				name,
				content,
				userId,
				guildId,
			})
			.returning();

		return [tag];
	});

	return tag;
}

export async function updateTag(
	name: string,
	guildId: string,
	values: PgUpdateSetSource<typeof schema.tags>,
) {
	const [updateTag] = await db.transaction(async (tx) => {
		const [updateGuild] = await tx
			.update(schema.tags)
			.set(values)
			.where(and(eq(schema.tags.name, name), eq(schema.tags.guildId, guildId)))
			.returning();

		return [updateGuild];
	});

	return updateTag;
}

File: akashi-database/src/models/tag.rs
use chrono::{NaiveDateTime, Utc};
use sqlx::{PgPool, Result};

type Snowflake = String;

#[derive(Debug, FromRow)]
pub struct Tag {
	pub guild_id: Snowflake,
	pub user_id: Snowflake,
	pub name: String,
	pub content: String,
	pub created_at: NaiveDateTime,
	pub views: i32,
}
impl Tag {
	/// Create a tag
	pub async fn create(
		pool: &PgPool,
		guild_id: String,
		user_id: String,
		name: String,
		content: String,
	) -> Result<Tag> {
		let tag = query_as!(
			Self,
			r#"
            INSERT INTO tags (guild_id, user_id, name, content, created_at, views)
            VALUES ($1, $2, $3, $4, $5, $6)
            RETURNING *
            "#,
			guild_id,
			user_id,
			name,
			content,
			Utc::now().naive_utc(),
			0
		)
		.fetch_one(pool)
		.await?;

		debug!("Created tag: {:?}", tag);

		Ok(tag)
	}

	/// Delete a tag by its name
	pub async fn delete(pool: &PgPool, guild_id: String, name: String) -> Result<()> {
		query!(
			r#"
            DELETE FROM tags
            WHERE guild_id = $1 AND name = $2
            "#,
			guild_id,
			name
		)
		.execute(pool)
		.await?;

		debug!("Deleted tag: {:?} in {:?}", name, guild_id);

		Ok(())
	}

	/// Get a tag by its name
	pub async fn get(pool: &PgPool, guild_id: String, name: String) -> Result<Option<Tag>> {
		let tag = query_as!(
			Self,
			r#"
            SELECT *
            FROM tags
            WHERE guild_id = $1 AND name = $2
            "#,
			guild_id,
			name
		)
		.fetch_optional(pool)
		.await?;

		Ok(tag)
	}

	/// Edit a tag by its name
	pub async fn edit(
		pool: &PgPool,
		guild_id: String,
		name: String,
		content: String,
	) -> Result<Tag> {
		let tag = query_as!(
			Self,
			r#"
            UPDATE tags
            SET content = $1
            WHERE guild_id = $2 AND name = $3
            RETURNING *
            "#,
			content,
			guild_id,
			name
		)
		.fetch_one(pool)
		.await?;

		debug!("Edited tag: {:?}", tag);

		Ok(tag)
	}

	/// Get all tags in a guild
	pub async fn list_guild(pool: &PgPool, guild_id: String) -> Result<Vec<Tag>> {
		let tags = query_as!(
			Tag,
			r#"
            SELECT *
            FROM tags
            WHERE guild_id = $1
            "#,
			guild_id
		)
		.fetch_all(pool)
		.await?;

		Ok(tags)
	}

	/// Get all tags in a guild with a user
	pub async fn list_user(pool: &PgPool, guild_id: String, user_id: String) -> Result<Vec<Tag>> {
		let tags = query_as!(
			Tag,
			r#"
            SELECT *
            FROM tags
            WHERE guild_id = $1 AND user_id = $2
            "#,
			guild_id,
			user_id
		)
		.fetch_all(pool)
		.await?;

		Ok(tags)
	}

	/// Increment the views of a tag
	pub async fn increment_views(pool: &PgPool, guild_id: String, name: String) -> Result<()> {
		query_as!(
			Tag,
			r#"
            UPDATE tags
            SET views = views + 1
            WHERE guild_id = $1 AND name = $2
            "#,
			guild_id,
			name
		)
		.execute(pool)
		.await?;

		Ok(())
	}
}
